<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>scRNAseq Data · SeqSpace.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><script src="../../../copy.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">SeqSpace.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Science</span><ul><li><a class="tocitem" href="../../sci/normalize/">scRNAseq Normalization</a></li><li><a class="tocitem" href="../../sci/inference/">scRNAseq Spatial Inference</a></li><li><a class="tocitem" href="../../sci/autoencode/">Manifold learning</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../distance/">Distances</a></li><li><a class="tocitem" href="../generate/">Point Cloud Generation</a></li><li><a class="tocitem" href="../infer/">Supervised Spatial Inference</a></li><li><a class="tocitem" href="../io/">Data Input/Output</a></li><li><a class="tocitem" href="../manifold/">Point Cloud Differential Geometry</a></li><li><a class="tocitem" href="../model/">Autoencoder Manifold Learning</a></li><li><a class="tocitem" href="../normalize/">scRNAseq Normalization</a></li><li><a class="tocitem" href="../pointcloud/">Point Cloud</a></li><li><a class="tocitem" href="../queue/">Priority Queue</a></li><li><a class="tocitem" href="../rank/">Differentiable Rank</a></li><li class="is-active"><a class="tocitem" href>scRNAseq Data</a><ul class="internal"><li><a class="tocitem" href="#Types"><span>Types</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li><li><a class="tocitem" href="../util/">Utilities</a></li><li><a class="tocitem" href="../voronoi/">Voronoi</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/drosophila/">Drosophila pipeline</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>scRNAseq Data</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>scRNAseq Data</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com//blob/master/docs/src/lib/scrna.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="scRNAseq-Data"><a class="docs-heading-anchor" href="#scRNAseq-Data">scRNAseq Data</a><a id="scRNAseq-Data-1"></a><a class="docs-heading-anchor-permalink" href="#scRNAseq-Data" title="Permalink"></a></h1><h2 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.scRNA.Count" href="#SeqSpace.scRNA.Count"><code>SeqSpace.scRNA.Count</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">struct Count{T &lt;: Real} &lt;: AbstractArray{T,2}
    data :: Array{T,2}
    gene :: Array{AbstractString,1}
    cell :: Array{AbstractString,1}
end</code></pre><p>Data structure used to represent count data obtained during a scRNAseq sequencing experiment. Individual cells are stored as column vectors while expression of singular genes are obtained as row vectors. <code>data</code> contains the raw/normalized count matrix. <code>gene</code> and <code>cell</code> contain the row/column labels respectively. Genes and cells can be indexed <em>either</em> by integers or names, i.e. strings.</p></div></section></article><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="Base.:∪-Union{Tuple{S}, Tuple{T}, Tuple{SeqSpace.scRNA.Count{T}, SeqSpace.scRNA.Count{S}}} where {T&lt;:Real, S&lt;:Real}" href="#Base.:∪-Union{Tuple{S}, Tuple{T}, Tuple{SeqSpace.scRNA.Count{T}, SeqSpace.scRNA.Count{S}}} where {T&lt;:Real, S&lt;:Real}"><code>Base.:∪</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">∪(seq₁::Count{T}, seq₂::Count{S}) where {T &lt;: Real, S &lt;: Real}</code></pre><p>Collate count matrix <code>seq₁</code> and <code>seq₂</code> by taking the union across genes. Reorders rows of <code>seq₂</code> to match gene names of <code>seq₁</code>. Additional genes in <code>seq₂</code> not contained in <code>seq₁</code> are added as augmented rows.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.scRNA.:∩-Union{Tuple{S}, Tuple{T}, Tuple{SeqSpace.scRNA.Count{T}, SeqSpace.scRNA.Count{S}}} where {T&lt;:Real, S&lt;:Real}" href="#SeqSpace.scRNA.:∩-Union{Tuple{S}, Tuple{T}, Tuple{SeqSpace.scRNA.Count{T}, SeqSpace.scRNA.Count{S}}} where {T&lt;:Real, S&lt;:Real}"><code>SeqSpace.scRNA.:∩</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">∩(seq₁::Count{T}, seq₂::Count{S}) where {T &lt;: Real, S &lt;: Real}</code></pre><p>Collate count matrix <code>seq₁</code> and <code>seq₂</code> by taking the union across genes. Reorders rows of <code>seq₂</code> to match gene names of <code>seq₁</code>. Only keeps genes present in <em>both</em> <code>seq₁</code> and <code>seq₂</code>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.scRNA.filtercell-Tuple{Any, SeqSpace.scRNA.Count}" href="#SeqSpace.scRNA.filtercell-Tuple{Any, SeqSpace.scRNA.Count}"><code>SeqSpace.scRNA.filtercell</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">filtercells(f, seq::Count)</code></pre><p>Filters cells of count matrix <code>seq</code> based upon column function <code>f</code>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.scRNA.filtergene-Tuple{Any, SeqSpace.scRNA.Count}" href="#SeqSpace.scRNA.filtergene-Tuple{Any, SeqSpace.scRNA.Count}"><code>SeqSpace.scRNA.filtergene</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">filtergene(f, seq::Count)</code></pre><p>Filters genes of count matrix <code>seq</code> based upon row function <code>f</code>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.scRNA.generate-Tuple{Any, Any}" href="#SeqSpace.scRNA.generate-Tuple{Any, Any}"><code>SeqSpace.scRNA.generate</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">generate(ngene, ncell; ρ=(α=Gamma(0.25,2), β=Normal(1,.01), γ=Gamma(3,3)))</code></pre><p>Generate scRNAseq data assuming a monoclonal population of cells sampled against a heteroskedastic negative binomial model.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.scRNA.load-Tuple{AbstractString}" href="#SeqSpace.scRNA.load-Tuple{AbstractString}"><code>SeqSpace.scRNA.load</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">load(dir::AbstractString; batch=missing)</code></pre><p>Read in scRNAseq experimental data from directory <code>dir</code>. The directory is expected to contain the following files:</p><ol><li>barcodes.tsv : one cell name per line</li><li>features.tsv : one gene name per line</li><li>matrix.mtx : count matrix in mtx format</li></ol><p>If <code>batch</code> is not missing, then it will be appended to each cell label.</p></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../rank/">« Differentiable Rank</a><a class="docs-footer-nextpage" href="../util/">Utilities »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.12 on <span class="colophon-date" title="Wednesday 23 February 2022 13:32">Wednesday 23 February 2022</span>. Using Julia version 1.7.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
