<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>scRNAseq Normalization Â· SeqSpace.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><script src="../../../copy.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">SeqSpace.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Science</span><ul><li><a class="tocitem" href="../../sci/normalize/">scRNAseq Normalization</a></li><li><a class="tocitem" href="../../sci/inference/">scRNAseq Spatial Inference</a></li><li><a class="tocitem" href="../../sci/autoencode/">Manifold learning</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../distance/">Distances</a></li><li><a class="tocitem" href="../generate/">Point Cloud Generation</a></li><li><a class="tocitem" href="../infer/">Supervised Spatial Inference</a></li><li><a class="tocitem" href="../io/">Data Input/Output</a></li><li><a class="tocitem" href="../manifold/">Point Cloud Differential Geometry</a></li><li><a class="tocitem" href="../model/">Autoencoder Manifold Learning</a></li><li class="is-active"><a class="tocitem" href>scRNAseq Normalization</a><ul class="internal"><li><a class="tocitem" href="#Types"><span>Types</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li><li><a class="tocitem" href="../pointcloud/">Point Cloud</a></li><li><a class="tocitem" href="../queue/">Priority Queue</a></li><li><a class="tocitem" href="../rank/">Differentiable Rank</a></li><li><a class="tocitem" href="../scrna/">scRNAseq Data</a></li><li><a class="tocitem" href="../util/">Utilities</a></li><li><a class="tocitem" href="../voronoi/">Voronoi</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/drosophila/">Drosophila pipeline</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>scRNAseq Normalization</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>scRNAseq Normalization</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com//blob/master/docs/src/lib/normalize.md" title="Edit on GitHub"><span class="docs-icon fab">ï‚›</span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="scRNAseq-Normalization"><a class="docs-heading-anchor" href="#scRNAseq-Normalization">scRNAseq Normalization</a><a id="scRNAseq-Normalization-1"></a><a class="docs-heading-anchor-permalink" href="#scRNAseq-Normalization" title="Permalink"></a></h1><h2 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Normalize.FitType" href="#SeqSpace.Normalize.FitType"><code>SeqSpace.Normalize.FitType</code></a> â€” <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">FitType = NamedTuple{
(
 :likelihood,
 :parameters,
 :uncertainty,
 :residual
),
Tuple{Float64, Vector{Float64}, Vector{Float64}, Vector{Float64}}</code></pre><p>Stores the result of MLE fit of <em>one</em> gene.</p></div></section></article><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Normalize.bootstrap-Tuple{Any, Any}" href="#SeqSpace.Normalize.bootstrap-Tuple{Any, Any}"><code>SeqSpace.Normalize.bootstrap</code></a> â€” <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">bootstrap(count, depth; stochastic=negativebinomial, samples=50)</code></pre><p>Empirically verify the MLE fit of <code>count</code>, using a GLM model generated by <code>stochastic</code> with confounding <code>depth</code> variables by bootstrap. One third of cells are removed and the parameters are re-estimated with the remaining cells. This process is repeated <code>samples</code> times. The resultant distribution of estimation is returned.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Normalize.fit-Tuple{Any, Any, Any}" href="#SeqSpace.Normalize.fit-Tuple{Any, Any, Any}"><code>SeqSpace.Normalize.fit</code></a> â€” <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">fit(stochastic, count, depth)</code></pre><p>Fit a generative model <code>stochastic</code> to gene expression <code>count</code> data, assuming confounding sequencing <code>depth</code>. <code>stochastic</code> can be either <a href="#SeqSpace.Normalize.negativebinomial-Tuple{Any, Any}"><code>negativebinomial</code></a> or <a href="#SeqSpace.Normalize.gamma-Tuple{Any, Any}"><code>gamma</code></a>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Normalize.gamma-Tuple{Any, Any}" href="#SeqSpace.Normalize.gamma-Tuple{Any, Any}"><code>SeqSpace.Normalize.gamma</code></a> â€” <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">gamma(count, depth)</code></pre><p>Compute the log likelihood of a gamma distributed generalized linear model (GLM) with log link function for the estimated mean <code>count</code> of a single gene. The sequencing depth for each sequenced cell is assumed to be the only confounding variables.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Normalize.glm-Tuple{Any}" href="#SeqSpace.Normalize.glm-Tuple{Any}"><code>SeqSpace.Normalize.glm</code></a> â€” <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">glm(data; stochastic=negativebinomial, Ïµ=1)</code></pre><p>Fit a generalized linear model (GLM) to the matrix <code>data</code>. Genes are assumed to be on rows, cells over columns. The underlying generative model is passed by <code>stochastic</code>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Normalize.logmean" href="#SeqSpace.Normalize.logmean"><code>SeqSpace.Normalize.logmean</code></a> â€” <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">logmean(x;Ïµ=1)</code></pre><p>Compute the geometric mean: <span>$xp\left(\langle \log\left(x + \epsilon \right) \rangle\right) - \epsilon$</span></p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Normalize.logvar" href="#SeqSpace.Normalize.logvar"><code>SeqSpace.Normalize.logvar</code></a> â€” <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">logmean(x;Ïµ=1)</code></pre><p>Compute the geometric variance: <span>$exp\left(\langle \log\left(x + \epsilon \right)^2 \rangle_c\right) - \epsilon$</span></p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Normalize.negativebinomial-Tuple{Any, Any}" href="#SeqSpace.Normalize.negativebinomial-Tuple{Any, Any}"><code>SeqSpace.Normalize.negativebinomial</code></a> â€” <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">negativebinomial(count, depth)</code></pre><p>Compute the log likelihood of a negative binomial generalized linear model (GLM) with log link function for <code>count</code> of a single gene. The sequencing depth for each sequenced cell is assumed to be the only confounding variables.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Normalize.prior-Tuple{Any}" href="#SeqSpace.Normalize.prior-Tuple{Any}"><code>SeqSpace.Normalize.prior</code></a> â€” <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">prior(params)</code></pre><p>Estimate a generalized normal distribution to <code>params</code> by maximum likelihood estimation. In practice, used to compute the empirical prior for overdispersion factor <code>Î˜â‚ƒ</code> in the negative binomial.</p></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../model/">Â« Autoencoder Manifold Learning</a><a class="docs-footer-nextpage" href="../pointcloud/">Point Cloud Â»</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.12 on <span class="colophon-date" title="Wednesday 23 February 2022 13:32">Wednesday 23 February 2022</span>. Using Julia version 1.7.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
