<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>scRNAseq Spatial Inference · SeqSpace.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><script src="../../../copy.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">SeqSpace.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Science</span><ul><li><a class="tocitem" href="../normalize/">scRNAseq Normalization</a></li><li class="is-active"><a class="tocitem" href>scRNAseq Spatial Inference</a><ul class="internal"><li><a class="tocitem" href="#Introduction"><span>Introduction</span></a></li><li><a class="tocitem" href="#Overview-of-current-methods"><span>Overview of current methods</span></a></li><li><a class="tocitem" href="#Our-approach"><span>Our approach</span></a></li><li><a class="tocitem" href="#Results"><span>Results</span></a></li><li><a class="tocitem" href="#Discussion"><span>Discussion</span></a></li></ul></li><li><a class="tocitem" href="../autoencode/">Manifold learning</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/distance/">Distances</a></li><li><a class="tocitem" href="../../lib/generate/">Point Cloud Generation</a></li><li><a class="tocitem" href="../../lib/infer/">Supervised Spatial Inference</a></li><li><a class="tocitem" href="../../lib/io/">Data Input/Output</a></li><li><a class="tocitem" href="../../lib/manifold/">Point Cloud Differential Geometry</a></li><li><a class="tocitem" href="../../lib/model/">Autoencoder Manifold Learning</a></li><li><a class="tocitem" href="../../lib/normalize/">scRNAseq Normalization</a></li><li><a class="tocitem" href="../../lib/pointcloud/">Point Cloud</a></li><li><a class="tocitem" href="../../lib/queue/">Priority Queue</a></li><li><a class="tocitem" href="../../lib/rank/">Differentiable Rank</a></li><li><a class="tocitem" href="../../lib/scrna/">scRNAseq Data</a></li><li><a class="tocitem" href="../../lib/util/">Utilities</a></li><li><a class="tocitem" href="../../lib/voronoi/">Voronoi</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/drosophila/">Drosophila pipeline</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Science</a></li><li class="is-active"><a href>scRNAseq Spatial Inference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>scRNAseq Spatial Inference</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com//blob/master/docs/src/sci/inference.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="scRNAseq-Spatial-Inference"><a class="docs-heading-anchor" href="#scRNAseq-Spatial-Inference">scRNAseq Spatial Inference</a><a id="scRNAseq-Spatial-Inference-1"></a><a class="docs-heading-anchor-permalink" href="#scRNAseq-Spatial-Inference" title="Permalink"></a></h1><h2 id="Introduction"><a class="docs-heading-anchor" href="#Introduction">Introduction</a><a id="Introduction-1"></a><a class="docs-heading-anchor-permalink" href="#Introduction" title="Permalink"></a></h2><p>In order to understand the regulation of morphogenesis, it is paramount to understand both the dynamics of gene expression at the single-cell level, as well as, interactions between the transcriptomic state of neighboring cells. With the advent of single-cell sequencing technology, it has now become possible to directly measure the transcriptome of cellular aggregates at cellular resolution, however, parameterizing such data by <em>space</em> remains challenging. Specifically, in the process of isolating individual cells for downstream sequencing, embryonic cells must be dissociated and suspended in liquid. This prevents straightforward application of scRNAseq technology to the problems of Developmental Biology.</p><p>The straightforward resolution to this conundrum is to leverage pre-existing databases of known in-situ markers. Such curated datasets should provide the spatial expression profiles for a small subset of genes that could be matched against scRNAseq counts measured from the same organism at the same stage of development. Once such a subset of genes are &quot;matched&quot; between the in-situ and scRNAseq data, the spatial pattern for each gene for the remainder of the transcriptome would come for &quot;free.&quot; The collection of estimated gene expression patterns would function as a high-resolution atlas over physical space of the embryo&#39;s shape; a potentially important resource for researchers.</p><p>Additionally, such an inference would provide us with an estimate of the position on the embryo each cell was sampled from. Assuming one has the ability to detect the intrinsic manifold of gene expression, as detailed elsewhere, such positional labels would provide an interesting overlay to attempt to learn the genotype to space map encoded by the genome. Below we provide an (incomplete) overview of past attempts at this problem, as well as detail our attempt to infer the position of scRNAseq cells.</p><h2 id="Overview-of-current-methods"><a class="docs-heading-anchor" href="#Overview-of-current-methods">Overview of current methods</a><a id="Overview-of-current-methods-1"></a><a class="docs-heading-anchor-permalink" href="#Overview-of-current-methods" title="Permalink"></a></h2><p><strong>FILL ME OUT</strong> There are published techniques that attempt to solve this problem, however our attempts to utilize them were unsuccessful.</p><h2 id="Our-approach"><a class="docs-heading-anchor" href="#Our-approach">Our approach</a><a id="Our-approach-1"></a><a class="docs-heading-anchor-permalink" href="#Our-approach" title="Permalink"></a></h2><h3 id="Objective-function"><a class="docs-heading-anchor" href="#Objective-function">Objective function</a><a id="Objective-function-1"></a><a class="docs-heading-anchor-permalink" href="#Objective-function" title="Permalink"></a></h3><p>Let <span>$\alpha$</span> and <span>$a$</span> denote indices over scRNAseq cells and embryonic spatial positions. Our goal is to solve for the sampling probability distribution <span>$\rho_{a\alpha}$</span>, i.e. the probability that cell <span>$\alpha$</span> was sampled from position <span>$a$</span>. We pose the language of regularized optimal transport, and thus the solution found at the extrema of the following objective function</p><p class="math-container">\[    \tag{1} F(\vec{g}_\alpha, \vec{G}_a) \equiv \displaystyle\sum\limits_{\alpha,a} C_{a\alpha}\left(\vec{g}_\alpha,\vec{G}_a\right)\rho_{a\alpha} + T\rho_{a\alpha}\log\left(\rho_{a\alpha}\right)\]</p><p>where <span>$\vec{g}_\alpha$</span> and <span>$\vec{G}_a$</span> denote the transcriptomic state of cell <span>$\alpha$</span> and position <span>$a$</span> respectively. In the parlance of optimal transport, <span>$C_{a\alpha}$</span> is a cost matrix - it denotes the energy required to map cell \alpha onto position a. For now, we leave the functional form general but understood to implicitly depend upon the gene expression of both sequenced cell and in-situ position. <span>$T$</span> is a hyperparameter akin to thermodynamic &quot;temperature&quot;; it controls the precision that we demand of the inferred position for each sequenced cell. As <span>$T \rightarrow 0$</span>, each cell must bijectively map onto a spatial position; the problem reduces to the assignment problem Conversely, as <span>$T \rightarrow \infty$</span>, the entropic term dominates; the solution is the uniform distribution.</p><p>If extremized as written, Eq. (1) would not result in a well-formed probability distribution. Specifically, we must constrain the row and column sum of <span>$\rho_{a\alpha}$</span> to have correctly interpreted marginals. In order for <span>$\rho_{a\alpha}$</span> to be interpreted as the probability that cell <span>$\alpha$</span> was sampled from position <span>$a$</span>, <span>$\forall \alpha \ \sum_a \rho_{a\alpha} = 1$</span> must hold. Additionally, we assume there were no biases in cellular isolation and thus each position was sequencing uniformally, and thus impose uniform coverage <span>$\forall a \ \sum_\alpha \rho_{a\alpha} = \frac{N_c}{N_x}$</span>. <span>$N_x$</span> and <span>$N_c$</span> denote the number of in-situ positions and sequenced cells respectively. Taken together, our full Free Energy is of the form</p><p class="math-container">\[    \tilde{F}(\vec{g}_\alpha,\vec{G}_a)\!\equiv\!\displaystyle\sum\limits_{\alpha,a} C_{a\alpha}\left(\vec{g}_\alpha,\vec{G}_a\right)\rho_{a\alpha} + T\rho_{a\alpha}\log\left(\rho_{a\alpha}\right) + \displaystyle\sum_a\Lambda_a\left[\frac{N_c}{N_x}\!-\!\sum_\alpha \rho_{a\alpha} \right] + \displaystyle\sum_\alpha \lambda_\alpha\left[1\!-\!\sum_a\rho_{a\alpha} \right]\]</p><p>The solution is found to be</p><p class="math-container">\[    \tag{2} \rho_{a\alpha}^* = e^{\lambda_a} e^{-T^{-1}\left(C_{a\alpha}\left(\vec{g}_\alpha,\vec{G}_a\right)-1\right)} e^{\Lambda_\alpha}\]</p><p>where <span>$\lambda_a$</span> and <span>$\Lambda_\alpha$</span> can be found by utilizing the Sinkhorn-Knopp algorithm in conjunction with the marginal constraints prescribed above. Eq. (2) provides a fast, scalable algorithm to estimate the sampling posterior given a cost matrix <span>$C_{a\alpha}(\vec{g}_a,\vec{G}_\alpha)$</span>. All that remains is to formulate an explicit model.</p><h3 id="Cost-matrix"><a class="docs-heading-anchor" href="#Cost-matrix">Cost matrix</a><a id="Cost-matrix-1"></a><a class="docs-heading-anchor-permalink" href="#Cost-matrix" title="Permalink"></a></h3><p>As seen by Eq. (2), the cost matrix can be viewed as the energy of a Boltzmann distribution: <span>$C_{a\alpha}-1 \equiv E\left(\vec{g}_\alpha, \vec{G}_a\right)$</span> Hence, an obvious interpretation of <span>$E(\vec{g}_\alpha, \vec{G}_a)$</span> is as the negative log-likelihood that <span>$\vec{g}_\alpha$</span> and <span>$\vec{G}_a$</span> were sampled from the sample entity. Our first simplifying assumption is that genes within the database are statistically independent of each other and thus the energy is additive</p><p class="math-container">\[    E(\vec{g}_\alpha, \vec{G}_a) = \frac{1}{N_g}\displaystyle\sum\limits_{i=1}^{N_g} \varepsilon\left(g_{\alpha i}, G_{ai}\right)\]</p><p>where <span>$i$</span> indexes genes and <span>$\varepsilon$</span> denotes the single-body energetics. Thus the problem has been reduced to parameterizing the log-likelihood that <span>$g_\alpha$</span> and <span>$G_a$</span> where sampled from the sample underlying cell. However, the complication is that our scRNAseq data and the in-situ expression database are not directly relatable; both data are in manifestly different unit systems! Furthermore, there is no <em>a priori</em> obvious functional relationship one can use for regression to ultimately transform counts in one dataset to another.</p><p>Overview:</p><ul><li>Look at each CDF of individual gene.</li><li>Optimal transport of each 1D distribution: can be solved analytically.</li><li>Denote cdf of <span>$i^{th}$</span> gene of scRNA and in-situ by <span>$\phi_i$</span> and <span>$\Phi_i$</span> respectively</li><li>Provides us a unique map from one space to another that minimizes distortions in the observed distributions.</li><li>Assume Gaussian sampling probability.</li><li>Mean is given by the in-situ database.</li><li>Variance of Gaussian simply renormalizes the temperature of the original problem. Can unambiguously set to 1.</li><li>Result is</li></ul><p class="math-container">\[    \varepsilon\left(g_{\alpha i}, G_{ai}\right) \equiv \left(\Phi^{-1}_i\left(\phi_i\left(g_{\alpha i}\right)\right) - G_{ai}\right)^2\]</p><h2 id="Results"><a class="docs-heading-anchor" href="#Results">Results</a><a id="Results-1"></a><a class="docs-heading-anchor-permalink" href="#Results" title="Permalink"></a></h2><h2 id="Discussion"><a class="docs-heading-anchor" href="#Discussion">Discussion</a><a id="Discussion-1"></a><a class="docs-heading-anchor-permalink" href="#Discussion" title="Permalink"></a></h2><p>Unfortunately, such databases only exist for a select few model organisms and thus limit the applicability of this approach. Our hope is to leverage phenomenology gleaned from the correlations between expression and space for organisms with such a database that </p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../normalize/">« scRNAseq Normalization</a><a class="docs-footer-nextpage" href="../autoencode/">Manifold learning »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.12 on <span class="colophon-date" title="Wednesday 23 February 2022 13:32">Wednesday 23 February 2022</span>. Using Julia version 1.7.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
