<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>scRNAseq spatial inference · SeqSpace.jl</title><script data-outdated-warner src="../../seqspace/assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../seqspace/assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><script src="../../../copy.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../seqspace/assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../seqspace/assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../seqspace/assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">SeqSpace.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Science</span><ul><li><a class="tocitem" href="../normalize/">scRNAseq normalization</a></li><li class="is-active"><a class="tocitem" href>scRNAseq spatial inference</a><ul class="internal"><li><a class="tocitem" href="#Introduction"><span>Introduction</span></a></li><li><a class="tocitem" href="#Overview-of-current-methods"><span>Overview of current methods</span></a></li><li><a class="tocitem" href="#Our-approach"><span>Our approach</span></a></li><li><a class="tocitem" href="#Results"><span>Results</span></a></li><li><a class="tocitem" href="#Discussion"><span>Discussion</span></a></li></ul></li><li><a class="tocitem" href="../autoencode/">Manifold learning</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/distance/">Distances</a></li><li><a class="tocitem" href="../../lib/generate/">Point Cloud Generation</a></li><li><a class="tocitem" href="../../lib/infer/">Supervised Spatial Inference</a></li><li><a class="tocitem" href="../../lib/io/">Data Input/Output</a></li><li><a class="tocitem" href="../../lib/manifold/">Point Cloud Differential Geometry</a></li><li><a class="tocitem" href="../../lib/model/">Autoencoder Manifold Learning</a></li><li><a class="tocitem" href="../../lib/normalize/">scRNAseq Normalization</a></li><li><a class="tocitem" href="../../lib/pointcloud/">Point Cloud</a></li><li><a class="tocitem" href="../../lib/queue/">Priority Queue</a></li><li><a class="tocitem" href="../../lib/rank/">Differentiable Rank</a></li><li><a class="tocitem" href="../../lib/scrna/">scRNAseq Data</a></li><li><a class="tocitem" href="../../lib/util/">Utilities</a></li><li><a class="tocitem" href="../../lib/voronoi/">Voronoi</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/drosophila/">Drosophila pipeline</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Science</a></li><li class="is-active"><a href>scRNAseq spatial inference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>scRNAseq spatial inference</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com//blob/master/docs/src/sci/inference.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="scRNAseq-spatial-inference"><a class="docs-heading-anchor" href="#scRNAseq-spatial-inference">scRNAseq spatial inference</a><a id="scRNAseq-spatial-inference-1"></a><a class="docs-heading-anchor-permalink" href="#scRNAseq-spatial-inference" title="Permalink"></a></h1><h2 id="Introduction"><a class="docs-heading-anchor" href="#Introduction">Introduction</a><a id="Introduction-1"></a><a class="docs-heading-anchor-permalink" href="#Introduction" title="Permalink"></a></h2><p>In order to understand the regulation of morphogenesis, it is paramount to understand both the dynamics of gene expression at the single-cell level, as well as, interactions between the transcriptomic state of neighboring cells. With the advent of single-cell sequencing technology, it has now become possible to directly measure the transcriptome of cellular aggregates at cellular resolution, however, parameterizing such data by <em>space</em> remains challenging. Specifically, in the process of isolating individual cells for downstream sequencing, embryonic cells must be dissociated and suspended in liquid. This prevents straightforward application of <em>scRNAseq</em> technology to the problems of Developmental Biology.</p><p>The straightforward resolution to this conundrum is to leverage pre-existing databases of known in-situ markers. Such curated datasets should provide the spatial expression profiles for a small subset of genes that could be matched against <em>scRNAseq</em> counts measured from the same organism at the same stage of development. Once such a subset of genes are &quot;matched&quot; between the in-situ and <em>scRNAseq</em> data, the spatial pattern for each gene for the remainder of the transcriptome would come for &quot;free.&quot; The collection of estimated gene expression patterns would function as a high-resolution atlas over physical space of the embryo&#39;s shape; a potentially important resource for researchers.</p><p>Additionally, such an inference would provide us with an estimate of the position on the embryo each cell was sampled from. Assuming one has the ability to detect the intrinsic manifold of gene expression, as detailed elsewhere, such positional labels would provide an interesting overlay to attempt to learn the genotype to space map encoded by the genome. Below we provide an (incomplete) overview of past attempts at this problem, as well as detail our attempt to infer the position of <em>scRNAseq</em> cells.</p><h2 id="Overview-of-current-methods"><a class="docs-heading-anchor" href="#Overview-of-current-methods">Overview of current methods</a><a id="Overview-of-current-methods-1"></a><a class="docs-heading-anchor-permalink" href="#Overview-of-current-methods" title="Permalink"></a></h2><p><strong>FILL ME OUT</strong> There are published techniques that attempt to solve this problem, however our attempts to utilize them were unsuccessful.</p><h2 id="Our-approach"><a class="docs-heading-anchor" href="#Our-approach">Our approach</a><a id="Our-approach-1"></a><a class="docs-heading-anchor-permalink" href="#Our-approach" title="Permalink"></a></h2><p>While the below discussion is general, all results shown within the context of <em>Drosophila melongaster</em> and thus will be utilizing the Berkeley Drosophila Transcriptional Network Project database <sup class="footnote-reference"><a id="citeref-1" href="#footnote-1">[1]</a></sup>. Specifically, the BDTNP collaboration produced a &quot;virtual embryo&quot; pointcloud discretizing the spatial expression pattern of <span>$84$</span> genes. The virtual embryo was constructed by aligning florescent pointcloud data obtained by FISH of similarly staged embryos.</p><p align="center">
<figure>
  <img src="/seqspace/assets/drosophila/bdtnp.jpg" width="99%" />
  <figurecaption>
  Cartoon of the procedure used to generate the virtual embryo by the BDTNP.
  </figurecaption>
</figure>
</p><h3 id="Objective-function"><a class="docs-heading-anchor" href="#Objective-function">Objective function</a><a id="Objective-function-1"></a><a class="docs-heading-anchor-permalink" href="#Objective-function" title="Permalink"></a></h3><p>Let <span>$\alpha$</span> and <span>$a$</span> denote indices over <em>scRNAseq</em> cells and embryonic spatial positions respectively. Our goal is to solve for the sampling probability distribution <span>$\rho_{a\alpha}$</span>, i.e. the probability that cell <span>$\alpha$</span> was sampled from position <span>$a$</span>. We pose this problem using the tools of <em>Statistical Physics</em> and thus formulate the solution as the extrema of the following free energy</p><p class="math-container">\[    \tag{1} F(\rho_{a\alpha}) \equiv \displaystyle\sum\limits_{\alpha,a} \epsilon\left(\vec{g}_\alpha,\vec{G}_a\right)\rho_{a\alpha} + T\rho_{a\alpha}\log\left(\rho_{a\alpha}\right)\]</p><p>where <span>$\vec{g}_\alpha$</span> and <span>$\vec{G}_a$</span> denote the transcriptomic state of cell <span>$\alpha$</span> and position <span>$a$</span> respectively. We note that, as formulated, this problem is equivalent to regularized optimal transport<sup class="footnote-reference"><a id="citeref-2" href="#footnote-2">[2]</a></sup>. <span>$\epsilon$</span> denotes the energy required to map cell <span>$\alpha$</span> onto position <span>$a$</span>. For now, we leave the functional form generic but denote that it explicitly depends upon the gene expression value of both the sequenced cell <span>$\alpha$</span> and the in-situ position <span>$a$</span>. Additionally, <span>$T$</span> is a parameter analogous to a thermodynamic &quot;temperature&quot;; it controls the precision that we demand of the inferred position for each sequenced cell. As <span>$T \rightarrow 0$</span>, each cell is constrained to injectively map onto a spatial position; the problem reduces to the assignment problem<sup class="footnote-reference"><a id="citeref-3" href="#footnote-3">[3]</a></sup>. Conversely, as <span>$T \rightarrow \infty$</span>, the entropic term dominates; the solution converges onto the uniform distribution.</p><p>However, if extremized as written, Eq. (1) would not result in a well-formed probability distribution. Specifically, we must additionally constrain the row and column sum of <span>$\rho_{a\alpha}$</span> to have correctly interpretable marginals. As such, in order for <span>$\rho_{a\alpha}$</span> to be interpreted as the probability that cell <span>$\alpha$</span> was sampled from position <span>$a$</span>, <span>$\forall \alpha \ \sum_a \rho_{a\alpha} = 1$</span> must hold. Additionally, we assume there were no biases in cellular isolation during the sequencing procedure.  Thus each position is assumed to be sampled uniformally, and thus impose uniform spatial coverage <span>$\forall a \ \sum_\alpha \rho_{a\alpha} = \frac{N_c}{N_x}$</span>. <span>$N_x$</span> and <span>$N_c$</span> denote the number of in-situ positions and sequenced cells respectively. Taken together, our full <em>free energy</em> is of the form</p><p class="math-container">\[    \tilde{F}(\rho_{a\alpha})\!\equiv\!\displaystyle\sum\limits_{\alpha,a} \epsilon\left(\vec{g}_\alpha,\vec{G}_a\right)\rho_{a\alpha} + T\rho_{a\alpha}\log\left(\rho_{a\alpha}\right) + \displaystyle\sum_a\Lambda_a\left[\frac{N_c}{N_x}\!-\!\sum_\alpha \rho_{a\alpha} \right] + \displaystyle\sum_\alpha \lambda_\alpha\left[1\!-\!\sum_a\rho_{a\alpha} \right]\]</p><p>The solution is found to be</p><p class="math-container">\[    \tag{2} \rho_{a\alpha}^* = e^{\Lambda_a} e^{-T^{-1}\left(\epsilon\left(\vec{g}_\alpha,\vec{G}_a\right)-1\right)} e^{\lambda_\alpha}\]</p><p>where <span>$\Lambda_a$</span> and <span>$\lambda_\alpha$</span> are determined by utilizing the Sinkhorn-Knopp algorithm in conjunction with the marginal constraints prescribed above. Eq. (2) provides a fast, scalable algorithm to estimate the sampling posterior given any cost function <span>$\epsilon(\vec{g}_a,\vec{G}_\alpha)$</span>. All that remains is to formulate an explicit model.</p><h3 id="Microscopic-model"><a class="docs-heading-anchor" href="#Microscopic-model">Microscopic model</a><a id="Microscopic-model-1"></a><a class="docs-heading-anchor-permalink" href="#Microscopic-model" title="Permalink"></a></h3><p>As seen by Eq. (2), the cost function can be viewed as the energy of a Boltzmann distribution: <span>$E\left(\vec{g}_\alpha, \vec{G}_a\right) \equiv \left(\vec{g}_\alpha, \vec{G}_a\right)$</span> Hence, an obvious interpretation of <span>$E(\vec{g}_\alpha, \vec{G}_a)$</span> is as the negative log-likelihood that <span>$\vec{g}_\alpha$</span> and <span>$\vec{G}_a$</span> were sampled from the sample entity. Our first simplifying assumption is that genes within the database are statistically independent of each other and thus the log likelihood is additive</p><p class="math-container">\[    \tag{3} E(\vec{g}_\alpha, \vec{G}_a) = \frac{1}{N_g}\displaystyle\sum\limits_{i=1}^{N_g} \varepsilon\left(g_{\alpha i}, G_{ai}\right)\]</p><p>where <span>$i$</span> indexes genes and <span>$\varepsilon$</span> denotes the single-body energetics. Thus the problem has been reduced to parameterizing the log-likelihood that <span>$g_{\alpha i}$</span> and <span>$G_{ai}$</span> were sampled from the <em>same</em> underlying cell. However, the complication is that our <em>scRNAseq</em> data and the in-situ expression database are not directly relatable. Both datasets are in manifestly different unit systems, the <em>scRNAseq</em> data are expressed in <em>UMI</em> counts while the underlying database is ultimately florescent intensity collated from a myriad of FISH experiments.</p><p>We postulate that the &quot;true&quot; transformation that maps <em>scRNAseq</em> counts <span>$g_{\alpha i}$</span> to florescent intensity <span>$G_{ai}$</span> should minimize distortions between both observed distortions under the action of said map. This ultimately suggests we identify the putative transformation via minimizing the Wasserstein metric via optimal transport <sup class="footnote-reference"><a id="citeref-4" href="#footnote-4">[4]</a></sup>. It has been shown <sup class="footnote-reference"><a id="citeref-5" href="#footnote-5">[5]</a></sup> that strictly convex cost functions <span>$\varepsilon$</span> between <span>$1$</span>D distributions admit a unique optimal transport solution. Specifically, if we denote the cumulative density of <span>$g_{\alpha i}$</span> and <span>$G_{ai}$</span> by <span>$\phi_{i}$</span> and <span>$\Phi_{i}$</span> respectively, the minimizing transformation is given by <span>$\Phi_{i}^{-1} \circ \phi_{i}$</span>. We assume a Gaussian sampling probability with mean given by the BDTNP database such that our one-body energy takes the form</p><p class="math-container">\[    \tag{4} \varepsilon\left(g_{\alpha i}, G_{ai}\right) \equiv \left(\Phi^{-1}_i\left(\phi_i\left(g_{\alpha i}\right)\right) - G_{ai}\right)^2\]</p><p>Importantly, we have enforced that each gene has unit variance within its sampling distribution. Non-unit uniform variance would simply rescale our temperature parameter and thus can be ignored. Conversely, heterogeneous variances would effectively act as a additive weighting prefactor in the summation of Eq. (3). We do not consider this case here but note that it is an interesting avenue for future improvement.</p><p>Eqns. (2-4) uniquely determine the sampling probability <span>$\rho_{a\alpha}$</span> modulo one free parameter, temperature <span>$T$</span>. <em>A priori</em> we expect the fit of gene expression to the database to be non-monotonic with respect to temperature. Minimizing of Eqn. (1) is singular as <span>$T \to 0$</span> and reduces to the assignment problem and thus is expected to be highly susceptible to noise. Conversely, as <span>$T \to \infty$</span> the entropic contribution to Eqn. (1) dominates and thus admits a uniform sampling distribution with no patterning. As such, we fix the temperature by hyperparameter optimization, i.e. to be the value that maximally correlates to the original BDTNP database.</p><h2 id="Results"><a class="docs-heading-anchor" href="#Results">Results</a><a id="Results-1"></a><a class="docs-heading-anchor-permalink" href="#Results" title="Permalink"></a></h2><p>Once the sampling probability <span>$\rho_{a\alpha}$</span> is known, we can immediately compute the mean expression profile for each gene</p><p class="math-container">\[    \bar{g}_{ia} \equiv \displaystyle\sum\limits_{\alpha} \rho_{a\alpha} g_{i\alpha}\]</p><p>We compare the computed mean expression profile to the &quot;known&quot; pattern as given by the BDTNP database. As shown in the figure below, we capture <span>$\sim 70\%$</span> of the variance of the database, on average across all <span>$84$</span> genes, at the optimal temperature. The red vertical bars display the standard deviation across all genes. The mapping at the optimal temperature corresponds to a sampling entropy <span>$\rho_{i\alpha}$</span> of <span>$\sim 6$</span> bits, corresponding to each scRNAseq cell mapping to <span>$\sim 60$</span> positions of the embryo.</p><p align="center">
<figure>
  <img src="/seqspace/assets/drosophila/bdtnp_fit.png" width="49%" />
  <img src="/seqspace/assets/drosophila/bdtnp_entropy.png" width="49%" />
  <figurecaption>
  (Left)Residuals of scRNAseq expression patterns matches to BDTNP database.
  (Right) Entropy of sampling distribution, averaged over cells.
  </figurecaption>
</figure>
</p><p>We note that our positional precision is significantly less than what has been found in previous studies utilizing florescence data <sup class="footnote-reference"><a id="citeref-6" href="#footnote-6">[6]</a></sup> that concluded gene expression could identify position with subcellular accuracy. The origin of the discovered &quot;imprecision&quot; is likely multifaceted owing, in part, to technical noise inherent to scRNAseq technology, systematic errors due to mistiming of the embryo stage between the database and scRNAseq data, as well as true biological variation associated to the full 2D positional information. Deconvolving the contributions of each stochastic process to our measurement error is an interesting future direction for research.</p><p>In addition to matching the small subset of genes contained within the BDTNP database, we also produce <em>predictions</em> for gene expression for the remaining <span>$\sim 10^4$</span> genes. The full database of predictions is available upon request; an exploratory web front-end is currently underway. Below we display a representative sample.</p><p align="center">
<figure>
  <img src="/seqspace/assets/drosophila/gene/eve.png" width="32%" />
  <img src="/seqspace/assets/drosophila/gene/ftz.png" width="32%" />
  <img src="/seqspace/assets/drosophila/gene/twi.png" width="32%" />
  <figurecaption>
  Genes shown (left-to-right): eve, ftz, twi
  </figurecaption>
</figure>
</p><h2 id="Discussion"><a class="docs-heading-anchor" href="#Discussion">Discussion</a><a id="Discussion-1"></a><a class="docs-heading-anchor-permalink" href="#Discussion" title="Permalink"></a></h2><p>Unfortunately, such databases only exist for a select few model organisms and thus limit the applicability of this approach. Our hope is to leverage phenomenology gleaned from the correlations between expression and space for organisms with such a database that </p><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-1"><a class="tag is-link" href="#citeref-1">1</a><a href="https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=1540644">Registering Drosophila Embryos at Cellular Resolution to Build a Quantitative 3D Atlas of Gene Expression Patterns and Morphology</a></li><li class="footnote" id="footnote-2"><a class="tag is-link" href="#citeref-2">2</a><a href="https://arxiv.org/abs/1306.0895">Sinkhorn Distances: Lightspeed Computation of Optimal Transportation Distances</a></li><li class="footnote" id="footnote-3"><a class="tag is-link" href="#citeref-3">3</a><a href="https://www.nowpublishers.com/article/DownloadSummary/MAL-073">Computational optimal transport: With applications to data science</a></li><li class="footnote" id="footnote-4"><a class="tag is-link" href="#citeref-4">4</a><a href="https://www.cedricvillani.org/sites/dev/files/old_images/2012/08/preprint-1.pdf">Optimal transport: old and new</a></li><li class="footnote" id="footnote-5"><a class="tag is-link" href="#citeref-5">5</a><a href="https://books.google.com/books/about/Mass_Transportation_Problems.html?id=t1LsSrWKjKQC">Mass Transportation Problems: Volume I: Theory</a></li><li class="footnote" id="footnote-6"><a class="tag is-link" href="#citeref-6">6</a><a href="https://www.pnas.org/content/110/41/16301">Positional information, in bits</a></li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../normalize/">« scRNAseq normalization</a><a class="docs-footer-nextpage" href="../autoencode/">Manifold learning »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.12 on <span class="colophon-date" title="Tuesday 1 March 2022 15:54">Tuesday 1 March 2022</span>. Using Julia version 1.7.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
