<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Supervised Spatial Inference · SeqSpace.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><script src="../../../copy.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">SeqSpace.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Science</span><ul><li><a class="tocitem" href="../../sci/normalize/">scRNAseq normalization</a></li><li><a class="tocitem" href="../../sci/inference/">scRNAseq spatial inference</a></li><li><a class="tocitem" href="../../sci/autoencode/">Manifold learning</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../distance/">Distances</a></li><li><a class="tocitem" href="../generate/">Point Cloud Generation</a></li><li class="is-active"><a class="tocitem" href>Supervised Spatial Inference</a><ul class="internal"><li><a class="tocitem" href="#Types"><span>Types</span></a></li><li><a class="tocitem" href="#Functions"><span>Functions</span></a></li></ul></li><li><a class="tocitem" href="../io/">Data Input/Output</a></li><li><a class="tocitem" href="../manifold/">Point Cloud Differential Geometry</a></li><li><a class="tocitem" href="../model/">Autoencoder Manifold Learning</a></li><li><a class="tocitem" href="../normalize/">scRNAseq Normalization</a></li><li><a class="tocitem" href="../pointcloud/">Point Cloud</a></li><li><a class="tocitem" href="../queue/">Priority Queue</a></li><li><a class="tocitem" href="../rank/">Differentiable Rank</a></li><li><a class="tocitem" href="../scrna/">scRNAseq Data</a></li><li><a class="tocitem" href="../util/">Utilities</a></li><li><a class="tocitem" href="../voronoi/">Voronoi</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/drosophila/">Drosophila pipeline</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Supervised Spatial Inference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Supervised Spatial Inference</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com//blob/master/docs/src/lib/infer.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Supervised-Spatial-Inference"><a class="docs-heading-anchor" href="#Supervised-Spatial-Inference">Supervised Spatial Inference</a><a id="Supervised-Spatial-Inference-1"></a><a class="docs-heading-anchor-permalink" href="#Supervised-Spatial-Inference" title="Permalink"></a></h1><h2 id="Types"><a class="docs-heading-anchor" href="#Types">Types</a><a id="Types-1"></a><a class="docs-heading-anchor-permalink" href="#Types" title="Permalink"></a></h2><h2 id="Functions"><a class="docs-heading-anchor" href="#Functions">Functions</a><a id="Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Functions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Inference.cost-Tuple{Any, Any}" href="#SeqSpace.Inference.cost-Tuple{Any, Any}"><code>SeqSpace.Inference.cost</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">cost(ref, qry; α=1, β=1, γ=0, ω=nothing)</code></pre><p>Return the cost matrix <span>$J_{ilpha}$</span> associated to matching cells in <code>qry</code> to cells in <code>ref</code>. The cost matrix is computed by a heuristic distance between quantiles. Deprecated.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Inference.cost_simple-Tuple{Any, Any}" href="#SeqSpace.Inference.cost_simple-Tuple{Any, Any}"><code>SeqSpace.Inference.cost_simple</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">cost_simple(ref, qry)</code></pre><p>Return the cost matrix <span>$J_{i\alpha}$</span> associated to matching cells in <code>qry</code> to cells in <code>ref</code>. The cost matrix is computed by hamming distance between cells via transforming quantiles to continuous spin variables. Deprecated.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Inference.cost_transform-Tuple{Any, Any}" href="#SeqSpace.Inference.cost_transform-Tuple{Any, Any}"><code>SeqSpace.Inference.cost_transform</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">cost_transform(ref, qry; ω=nothing, ν=nothing)</code></pre><p>Return the cost matrix <span>$J_{i\alpha}$</span> associated to matching cells in <code>qry</code> to cells in <code>ref</code>. The cost matrix is computed by:</p><ol><li>Transforming the <code>qry</code> distribution to the <code>ref</code> distribution.</li><li>Looking at the SSE across transformed genes.</li></ol><p>Use this unless you know what you are doing.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Inference.inversion-Tuple{Any, Any}" href="#SeqSpace.Inference.inversion-Tuple{Any, Any}"><code>SeqSpace.Inference.inversion</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">inversion(counts, genes; ν=nothing, ω=nothing, refdb=nothing)</code></pre><p>Infer the original position of scRNAseq data <code>counts</code> where genes, given by <code>genes</code> are arranged along rows. The sampling probability over space is computed by regularized optimal transport by comparing to the Berkeley Drosophila Transcription Network Project database. The cost matrix is determined by summing over the 1D Wasserstein metric over all genes within the BDTNP databse. Returns the inversion as a function of inverse temperature.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Inference.sinkhorn-Tuple{Matrix{Float64}}" href="#SeqSpace.Inference.sinkhorn-Tuple{Matrix{Float64}}"><code>SeqSpace.Inference.sinkhorn</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">sinkhorn(M::Array{Float64,2};
              a::Maybe{Array{Float64}} = missing,
              b::Maybe{Array{Float64}} = missing,
              maxᵢ::Integer            = 1000,
              τ::Real                  = 1e-5,
              verbose::Bool            = false
)</code></pre><p>Rescale matrix <code>M</code> to have row &amp; column marginals <code>a</code> and <code>b</code> respectively. Will terminate either when constraints are held to within tolerance <code>τ</code> or the number of iterations exceed <code>maxᵢ</code>.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Inference.transform-Tuple{Any, Any, Any}" href="#SeqSpace.Inference.transform-Tuple{Any, Any, Any}"><code>SeqSpace.Inference.transform</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">transform(src, dst, ν)</code></pre><p>Transform distribution <code>src</code> to distribution <code>dst</code> by minimizing the Wasserstein metric. This is equivalent to <span>$x \to F^{-1}_{dst}\left(F_{src}\left(x\right)\right)$</span> where <span>$F$</span> denotes the cumulative density function.</p></div></section></article><article class="docstring"><header><a class="docstring-binding" id="SeqSpace.Inference.virtualembryo-Tuple{}" href="#SeqSpace.Inference.virtualembryo-Tuple{}"><code>SeqSpace.Inference.virtualembryo</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">virtualembryo(;directory=&quot;/home/nolln/mnt/data/drosophila/dvex&quot;)</code></pre><p>Load the Berkeley Drosophila Transcriptional Network Project database. <code>directory</code> should be path to folder containing two folders:</p><ol><li>bdtnp.txt.gz    : gene expression over point cloud of virtual cells</li><li>geometry.txt.gz : spatial position (x,y,z) of point cloud of virtual cells.</li></ol></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../generate/">« Point Cloud Generation</a><a class="docs-footer-nextpage" href="../io/">Data Input/Output »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.12 on <span class="colophon-date" title="Tuesday 1 March 2022 15:54">Tuesday 1 March 2022</span>. Using Julia version 1.7.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
