<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>scRNAseq normalization · SeqSpace.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><script src="../../../copy.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">SeqSpace.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Science</span><ul><li class="is-active"><a class="tocitem" href>scRNAseq normalization</a><ul class="internal"><li><a class="tocitem" href="#Introduction"><span>Introduction</span></a></li><li><a class="tocitem" href="#Overview-of-current-methods"><span>Overview of current methods</span></a></li><li><a class="tocitem" href="#Our-approach"><span>Our approach</span></a></li><li><a class="tocitem" href="#Empirical-sampling-distribution"><span>Empirical sampling distribution</span></a></li><li><a class="tocitem" href="#Pearson-residuals"><span>Pearson residuals</span></a></li></ul></li><li><a class="tocitem" href="../inference/">scRNAseq spatial inference</a></li><li><a class="tocitem" href="../autoencode/">Manifold learning</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/distance/">Distances</a></li><li><a class="tocitem" href="../../lib/generate/">Point Cloud Generation</a></li><li><a class="tocitem" href="../../lib/infer/">Supervised Spatial Inference</a></li><li><a class="tocitem" href="../../lib/io/">Data Input/Output</a></li><li><a class="tocitem" href="../../lib/manifold/">Point Cloud Differential Geometry</a></li><li><a class="tocitem" href="../../lib/model/">Autoencoder Manifold Learning</a></li><li><a class="tocitem" href="../../lib/normalize/">scRNAseq Normalization</a></li><li><a class="tocitem" href="../../lib/pointcloud/">Point Cloud</a></li><li><a class="tocitem" href="../../lib/queue/">Priority Queue</a></li><li><a class="tocitem" href="../../lib/rank/">Differentiable Rank</a></li><li><a class="tocitem" href="../../lib/scrna/">scRNAseq Data</a></li><li><a class="tocitem" href="../../lib/util/">Utilities</a></li><li><a class="tocitem" href="../../lib/voronoi/">Voronoi</a></li></ul></li><li><span class="tocitem">Command Line</span><ul><li><a class="tocitem" href="../../cli/drosophila/">Drosophila pipeline</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Science</a></li><li class="is-active"><a href>scRNAseq normalization</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>scRNAseq normalization</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com//blob/master/docs/src/sci/normalize.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="scRNAseq-normalization"><a class="docs-heading-anchor" href="#scRNAseq-normalization">scRNAseq normalization</a><a id="scRNAseq-normalization-1"></a><a class="docs-heading-anchor-permalink" href="#scRNAseq-normalization" title="Permalink"></a></h1><h2 id="Introduction"><a class="docs-heading-anchor" href="#Introduction">Introduction</a><a id="Introduction-1"></a><a class="docs-heading-anchor-permalink" href="#Introduction" title="Permalink"></a></h2><p>scRNA sequencing is a relatively novel technique added to the Biologist&#39;s toolkit that allows one to quantitatively probe the &quot;Statistical Mechanics&quot; of Biology. In contrast to traditional bulk RNAseq techniques, which provide population averages, scRNAseq technology measures the complete ensemble of cellular gene expression. Consequently, considerable activity has been focused on mapping the taxonomy of cell states; projects such as the Human Cell Atlas promise to provide a complete enumeration of microscopic cell types within the human body. However, despite recent progress, scRNAseq data have many numerous technical limitations and sources of statistical bias that must be considered before any downstream analysis. The sources of noise within the data include, but are not limited to:</p><ul><li><strong>Sequence depth bias:</strong> PCR and reverse transcription efficiency will vary across reactions, resulting in artificial variation of sequencing depth across different cells.</li><li><strong>Amplification bias:</strong> PCR primers are not perfectly random. As a result, some genes will amplify preferentially over others. This will distort the underlying expression distribution.</li><li><strong>Batch effect:</strong> scRNAseq runs have non-trivial distortions causing cells within a given sequencing run to have greater correlation within than across batches.</li><li><strong>Dropout:</strong> Due to molecular competition in the underlying amplification reactions, some genes will fail to amplify during PCR due to early losses. This will lead to more zeros in the resulting count matrix than you would expect by chance.</li></ul><p>These biases are approximately rectified by a preprocessing step known as <strong>normalization</strong>. At a coarse level, all such methods can be thought of as transforming the obtained count matrix from absolute numbers into an estimate of differential expression, i.e. a comparison of a count <em>relative</em> to the measured distribution. Such a procedure is analogous to that of the z-score of a univariate normally-distributed random variable; however, in the case of scRNAseq data, we don&#39;t have access to the underlying sampling distributions <em>a priori</em>, it must be estimated empirically. The choice of sampling prior, and details of how the distribution is estimated, delineate normalization practices.</p><h2 id="Overview-of-current-methods"><a class="docs-heading-anchor" href="#Overview-of-current-methods">Overview of current methods</a><a id="Overview-of-current-methods-1"></a><a class="docs-heading-anchor-permalink" href="#Overview-of-current-methods" title="Permalink"></a></h2><p><strong>TODO:</strong> FILL OUT</p><h2 id="Our-approach"><a class="docs-heading-anchor" href="#Our-approach">Our approach</a><a id="Our-approach-1"></a><a class="docs-heading-anchor-permalink" href="#Our-approach" title="Permalink"></a></h2><p>We model the count matrix <span>$n_{\alpha,i}$</span>, where <span>$\alpha$</span> and <span>$i$</span> indexes cells and genes respectively, obtained from scRNA sequencing as an unknown, low-rank mean <span>$\mu_{\alpha,i}$</span> with additive full-rank noise <span>$\delta_{\alpha,i}$</span> that captures all unknown technical noise and bias.</p><p class="math-container">\[    \tag{1} n_{\alpha,i} = \mu_{\alpha,i} + \delta_{\alpha,i} \]</p><p>Our goal during the normalization procedure is two-fold:</p><ol><li>Estimate the low-rank mean <span>$\mu_{\alpha,i}$</span>.</li><li>Estimate the sampling variance <span>$\langle\delta_{\alpha,i}^2\rangle$</span> of the experimental noise. Brackets denote an average over (theoretical) realizations of sequencing. This will eventually require an explicit model.</li></ol><p align="center">
<figure>
  <img src="/seqspace/assets/drosophila/heteroskedastic.png" width="79%" />
  <figurecaption>
  Count matrix is heteroskedastic: variation of scales across genes (rows) and cells (columns)
  </figurecaption>
</figure>
</p><p>Given both quantities, normalization over gene <em>and</em> cell-specific biases can be achieved by imposing the variances of all marginal distributions to be one. Specifically, we rescale all counts by cell-specific <span>$c_\alpha$</span> and gene-specific <span>$g_i$</span> factors <span>$\tilde{n}_{\alpha,i} \equiv c_\alpha n_{\alpha,i} g_i$</span> such that</p><p class="math-container">\[    \tag{2} \displaystyle\sum\limits_{\alpha} \langle \tilde{\delta}_{\alpha,i}^2 \rangle = \displaystyle\sum\limits_{\alpha}c_\alpha^2 \langle \delta_{\alpha,i}^2 \rangle g_i^2 = N_g \quad \text{and} \quad \displaystyle\sum\limits_{i} \langle \tilde{\delta}_{\alpha,i}^2 \rangle = \displaystyle\sum\limits_{i}c_\alpha^2 \langle \delta_{\alpha,i}^2 \rangle g_i^2 = N_c\]</p><p>This system of equations can be solved by the Sinkhorn-Knopp algorithm, provided we have a model for <span>$\langle\delta_{\alpha,i}^2\rangle$</span> parameterized by measurables. Within this formalism, this choice of model fully determines the normalization scheme. Owing to dropout and other sources of overdispersion, as shown empircally in the figure below, we model scRNAseq counts as sampled from an empirically estimated <a href="#Heteroskedastic-Negative-Binomial">Heteroskedastic Negative Binomial</a> distribution.</p><p align="center">
<figure>
  <img src="/seqspace/assets/drosophila/overdispersed_mean_vs_variance.png" width="49%" />
  <img src="/seqspace/assets/drosophila/overdispersed_zeros.png" width="49%" />
  <figurecaption>
  scRNAseq data for Drosophila is overdispersed.
  </figurecaption>
</figure>
</p><h3 id="Case-studies"><a class="docs-heading-anchor" href="#Case-studies">Case studies</a><a id="Case-studies-1"></a><a class="docs-heading-anchor-permalink" href="#Case-studies" title="Permalink"></a></h3><p>Before detailing our explicit algorithm, it is helpful to consider a few simpler examples. In the following sections, homoskedastic is used to denote count matrices whose elements are  independent and identically distributed (<em>iid</em>), while heteroskedastic denotes more complicated scenarios where each element has a unique distribution.</p><h4 id="Homoskedastic-Gaussian"><a class="docs-heading-anchor" href="#Homoskedastic-Gaussian">Homoskedastic Gaussian</a><a id="Homoskedastic-Gaussian-1"></a><a class="docs-heading-anchor-permalink" href="#Homoskedastic-Gaussian" title="Permalink"></a></h4><p>This is slight perturbation of canonical <a href="https://en.wikipedia.org/wiki/Random_matrix">random matrix theory</a> and will serve as a useful pedagogical starting point. Assume <span>$\mu_{\alpha,i}$</span> is a quenched, low-rank matrix and each element of <span>$\delta_{\alpha,i}$</span> is sampled from a Gaussian with zero mean and variance <span>$\sigma^2$</span>. In this limit, Eq.(1) reduces to the well-studied spiked population covariance model.</p><p>If <span>$\mu_{\alpha,i} = 0$</span>, the spectral decomposition of the count matrix <span>$n_{\alpha,i}$</span> would be given by the <a href="https://en.wikipedia.org/wiki/Marchenko–Pastur_distribution">Marchenko-Pastur distribution</a> asymptotically. As such, singular values would be bounded by <span>$\bar{\lambda} \equiv 1+\sigma\sqrt{N_g/N_c}$</span>.</p><p align="center">
<img src="/seqspace/assets/drosophila/marchenko-pastur.png" width="49%" class="center"/>
</p>
<p align="center">
Marchenko pastur distribution (orange) vs random matrix eigenvalues (blue)
</p><p>Now consider the case of a rank 1 mean, i.e. 1 cell type in the data.</p><p class="math-container">\[    n_{\alpha,i} = \gamma x_\alpha \bar{x}_i + \delta_{\alpha,i}\]</p><p>It has been shown<sup class="footnote-reference"><a id="citeref-1" href="#footnote-1">[1]</a></sup><sup class="footnote-reference"><a id="citeref-2" href="#footnote-2">[2]</a></sup> that this model exhibits the following asymptotic phase transition:</p><ul><li>If <span>$\gamma \le \bar{\lambda}$</span>, the top singular value of <span>$n_{\alpha,i}$</span> converges to <span>$\bar{\lambda}$</span>. Additionally, the overlap of the left and right eigenvector with <span>$x$</span> and <span>$\bar{x}$</span> respectively converge to 0.</li><li>If <span>$\gamma &gt; \bar{\lambda}$</span>, the top singular value of <span>$n_{\alpha,i}$</span> converges to <span>$\gamma + \bar{\lambda}/\gamma$</span>. Additionally, the overlap of the left and right eigenvector with <span>$x$</span> and <span>$\bar{x}$</span> respectively converge to <span>$1-(\bar{\lambda}/\gamma)^2$</span></li></ul><p>This procedure can generalized to higher rank spike-ins; sub-leading principal components can be found by simply subtracting the previously inferred component from the count matrix <span>$n_{\alpha,i}$</span>. As such, we can only expect to meaningful measure the principal components of <span>$\mu_{\alpha,i}$</span> that fall above the sampling noise floor, given by the Marchenko-Pastur distribution. Consequently, this forces us to define the statistically significant <strong>rank</strong> of the count matrix. This is exhibited empirically below.</p><p align="center">
<img src="/seqspace/assets/drosophila/gaussian_svd.png" width="49%" class="center"/>
<img src="/seqspace/assets/drosophila/gaussian_overlap.png" width="49%" class="center"/>
</p><h4 id="Heteroskedastic-Poisson"><a class="docs-heading-anchor" href="#Heteroskedastic-Poisson">Heteroskedastic Poisson</a><a id="Heteroskedastic-Poisson-1"></a><a class="docs-heading-anchor-permalink" href="#Heteroskedastic-Poisson" title="Permalink"></a></h4><p>The above results have been shown to hold for non-Gaussian Wigner noise matrices<sup class="footnote-reference"><a id="citeref-3" href="#footnote-3">[3]</a></sup>, provided each element is still <em>iid</em>. This is manifestly <em>not</em> the case we care about; the normalization procedure must account for heterogeneous sampling variances across both cells and genes. However, it has been shown<sup class="footnote-reference"><a id="citeref-4" href="#footnote-4">[4]</a></sup><sup class="footnote-reference"><a id="citeref-5" href="#footnote-5">[5]</a></sup> that the distribution of eigenvalues converges almost surely to the Marchenko-Pastur distribution, provided the constraint of Eq.(2) is satisfied! In other words, the eigenvalues of a heteroskedastic matrix is expected converge to that of a random Wigner matrix, provided the row and column variances are uniform (set to unity for convenience).</p><p>For the remainder of this section, assume the count matrix is sampled from a Poisson distribution, such that <span>$\langle\delta_{\alpha,i}^2\rangle = \mu_{\alpha,i}$</span>. As <span>$n_{\alpha,i}$</span> is an unbiased estimator for the mean <span>$\mu_{\alpha,i}$</span>, Eq (2) reduces to</p><p class="math-container">\[\displaystyle\sum\limits_{\alpha}c_\alpha^2 n_{\alpha, i} g_i^2 = N_g \quad \text{and} \quad \displaystyle\sum\limits_{i}c_\alpha^2 n_{\alpha,i} g_i^2 = N_c\]</p><p>which provides an explicit system of equations to estimate the scaling factors <span>$c_\alpha$</span> and <span>$g_i$</span>. Once obtained, <span>$\tilde{\mu}_{\alpha,i}$</span> can be estimated via singular value decomposition of <span>$c_\alpha n_{\alpha,i} g_i$</span>; all components with singular value greater than <span>$\bar{\lambda} = \sqrt{N_c}+\sqrt{N_g}$</span> can be confidently attributed to the &quot;true mean&quot; <span>$\mu_{\alpha,i}$</span> while all other components fall amongst the noise. An example is shown below:</p><p align="center">
<img src="/seqspace/assets/drosophila/poisson_svd.png" width="49%" class="center"/>
<img src="/seqspace/assets/drosophila/poisson_overlap.png" width="49%" class="center"/>
</p><h4 id="Heteroskedastic-Negative-Binomial"><a class="docs-heading-anchor" href="#Heteroskedastic-Negative-Binomial">Heteroskedastic Negative Binomial</a><a id="Heteroskedastic-Negative-Binomial-1"></a><a class="docs-heading-anchor-permalink" href="#Heteroskedastic-Negative-Binomial" title="Permalink"></a></h4><p>A negative binomial distribution is often used to model overdispersed count data, i.e. a process in which the variance grows superlinearly with the mean. Canonically the distribution arises from the distribution of the number of successes (with probabiliy <span>$p$</span>) obtained after <span>$\Theta_3$</span> failures of a Bernoulli process. However, the generative stochastic process can equivalently be modelled as an underlying Poisson process in which the emission rate is itself a stochastic variable drawn from a Gamma distribution. This allows us to analytically continue <span>$\Theta_{3}$</span> from an integer to the reals. Provided we have estimated the mean <span>$\mu_{i\alpha}$</span> and the overdisperson factor <span>$\Theta_{3;i}$</span>, the unbiased estimator for the variance is given by</p><p class="math-container">\[    \tag{3} \langle \delta_{i\alpha}^2 \rangle = \mu_{i\alpha}\left(\frac{1 + \mu_{i\alpha}\Theta_{3;i}}{1 + \Theta_{3;i}}\right)\]</p><p>Direct substitution into Eq. (2) would provide the necessary cell-specific <span>$c_\alpha$</span> and gene-specific <span>$g_i$</span> scaling factors.</p><h2 id="Empirical-sampling-distribution"><a class="docs-heading-anchor" href="#Empirical-sampling-distribution">Empirical sampling distribution</a><a id="Empirical-sampling-distribution-1"></a><a class="docs-heading-anchor-permalink" href="#Empirical-sampling-distribution" title="Permalink"></a></h2><p>In order to normalize the estimated sampling variance, we must formulate a method to fit a negative binomial to the measured count data <em>per gene</em>. We parameterize the distribution as follows:</p><p class="math-container">\[    p\left(n|\mu,\Theta_{3}\right) = \frac{\Gamma\left(n+\Theta_{3}\right)}{\Gamma\left(n+1\right)\Gamma\left(\Theta_{3}\right)}\left(\frac{\mu}{\mu+\Theta_{3}}\right)^n\left(\frac{\Theta_{3}}{\mu+\Theta_{3}}\right)^{\Theta_{3}}\]</p><p>All results, unless explicitly stated otherwise, are obtained empirically from scRNAseq data obtained during early <em>Drosophila melongaster</em> embryogenesis.</p><h3 id="Generalized-linear-model"><a class="docs-heading-anchor" href="#Generalized-linear-model">Generalized linear model</a><a id="Generalized-linear-model-1"></a><a class="docs-heading-anchor-permalink" href="#Generalized-linear-model" title="Permalink"></a></h3><p>A central complication in modeling the counts of a given gene <em>across</em> cells with the above generative model is that there are confounding variables that require consideration. For the present discussion, we explictly model the hetereogeneous sequencing depth across cells: naively we expect that if we sequenced a given cell with <span>$2\times$</span> depth, each gene should scale accordingly. As such, we formulate a generalized linear model</p><p class="math-container">\[    \tag{4} \log \mu_{i\alpha} = \Theta_{1;i} + \Theta_{2;i} \log \chi_{\alpha}\]</p><p>where <span>$\chi_\alpha$</span> denotes the total sequencing depth of cell <span>$\alpha$</span> We note this formulation can be easily extended to account for other confounds such as batch effect or cell type. The likelihood function for cell <span>$\alpha$</span>, gene <span>$i$</span> is</p><p class="math-container">\[    p\left(n_{i\alpha}|\Theta_{1;i},\Theta_{2;i}, \Theta_{3;i},\chi_\alpha\right) = \frac{\Gamma\left(n_{i\alpha}+\Theta_{3;i}\right)}{\Gamma\left(n_{i\alpha}+1\right)\Gamma\left(\Theta_{3;i}\right)}\left(\frac{\mu_{i\alpha}}{\mu_{i\alpha}+\Theta_{3;i}}\right)^{n_{i\alpha}}\left(\frac{\Theta_{3;i}}{\mu_{i\alpha}+\Theta_{3;i}}\right)^{\Theta_{3;i}}\]</p><h3 id="Maximum-Likelihood-Estimation"><a class="docs-heading-anchor" href="#Maximum-Likelihood-Estimation">Maximum Likelihood Estimation</a><a id="Maximum-Likelihood-Estimation-1"></a><a class="docs-heading-anchor-permalink" href="#Maximum-Likelihood-Estimation" title="Permalink"></a></h3><p><span>$\{\Theta_{1;i}, \Theta_{2;i}, \Theta_{3;i} \}$</span> represent <span>$3N_g$</span> parameters we must infer from the data. We note that <em>a priori</em> this problem is overdetermined; our count matrix is sized <span>$N_g \times N_c$</span>  and thus we attempt to estimate all parameters within a maximum likelihood framework. This is equivalent to minimizing (for each gene independently)</p><p class="math-container">\[\begin{aligned}
    \tag{5} \mathcal{L}_i = \displaystyle\sum\limits_{\alpha}
        \left(n_{i\alpha} + \Theta_{3;i}\right)\log\left(e^{\Theta_{1;i}}\chi_\alpha^{\Theta_{2;i}} + \Theta_{3;i}\right) 
      - \Theta_{3;i}\log\left(\Theta_{3;i}\right) \\
      - n_{i\alpha}\left(\Theta_{1;i} + \Theta_{2;i}\log \chi_\alpha\right)
      - \log\left(\frac{\Gamma\left(n_{i\alpha}+\Theta_{3;i}\right)}{\Gamma\left(n_{i\alpha}+1\right)\Gamma\left(\Theta_{3;i}\right)}\right)
\end{aligned}\]</p><p>Empirically it was determined that to provide a robust estimate of all three parameters that our confounding variables <span>$\chi_{\alpha}$</span> are given by</p><p class="math-container">\[    \chi_\alpha \equiv \exp\left(\langle \log\left(n_{i\alpha}+1\right) \rangle\right) - 1\]</p><p>where angle brackets denote the empirical average over genes for cell <span>$\alpha$</span>. Once the parameters <span>$\{\Theta_{1;i}, \Theta_{2;i}, \Theta_{3;i} \}$</span> are estimated, Eqns. (2-4) can be utilized to estimate the normalized mean count matrix <span>$\tilde{\mu}_{i\alpha}$</span>. This will also immediately teach us the <strong>statistically significant rank</strong> of the counting matrix.</p><h4 id="Synthetic-data-verification"><a class="docs-heading-anchor" href="#Synthetic-data-verification">Synthetic data verification</a><a id="Synthetic-data-verification-1"></a><a class="docs-heading-anchor-permalink" href="#Synthetic-data-verification" title="Permalink"></a></h4><p>As a first check on the methodology, we generated toy counting data sampled from negative binomial distribution with low rank mean. After estimation of parameters by minimizing Eq. (5), we solved Eq. (2) utilizing our unbiased estimator given by Eq. (3). The result is shown below.</p><p align="center">
<img src="/seqspace/assets/drosophila/negbinom.png" width="49%" class="center"/>
<img src="/seqspace/assets/drosophila/negbinom_mean.png" width="49%" class="center"/>
</p><p>As shown, we underestimate the true rank as a few components have singular values below the Marchenko-Pastur noise floor. This, along with our noisy estimation of the overdispersion factor <span>$\Theta_3$</span> contributes to our noisier mean estimation when compared to the Poisson case above.</p><h4 id="Filter-uncertain-genes"><a class="docs-heading-anchor" href="#Filter-uncertain-genes">Filter uncertain genes</a><a id="Filter-uncertain-genes-1"></a><a class="docs-heading-anchor-permalink" href="#Filter-uncertain-genes" title="Permalink"></a></h4><p>The uncertainty of our parameter estimates can be computed by the second derivative of the likelihood around our determined minima</p><p class="math-container">\[    \delta \Theta_{a}^2 = \left[\partial_{\Theta_{b}}\partial_{\Theta_{c}} \mathcal{L}\right]^{-1}_{aa}\]</p><p>Unsurprisingly, the uncertainty of our estimates for parameters <span>$\{\Theta_{1;i}, \Theta_{2;i}, \Theta_{3;i} \}$</span> is strongly dependent upon the underlying gene expression; our estimates for lowly expressed genes are highly uncertain. This can be seen in the below figure, which shows the scatter plot of the parameter estimate <span>$\Theta_{a}$</span> versus its uncertainty <span>$\delta \Theta_a$</span>, colored by the average gene expression.</p><p align="center">
<figure>
  <img src="/seqspace/assets/drosophila/nb_1_uncertainty_vs_expression.png" width="32%" />
  <img src="/seqspace/assets/drosophila/nb_2_uncertainty_vs_expression.png" width="32%" />
  <img src="/seqspace/assets/drosophila/nb_3_uncertainty_vs_expression.png" width="32%" />
  <figurecaption>
  Each point is a gene (row). Color of point determined by mean expression of gene.
  </figurecaption>
</figure>
</p><p>We note that the estimated <span>$\Theta_1$</span> monotonically grows with increasing gene expression, as was expected by construction. Conversely, the average estimate for <span>$\Theta_2$</span> shows no obvious trend against expression levels, however the uncertainty <span>$\delta\Theta_2$</span> monotonically increases with decreasing gene expression. The estimated <span>$\Theta_3$</span> also appears to be independent of average expression counts with a family of curves with increasing uncertainty defined by decreasing gene levels.</p><p>We capture the <em>total</em> uncertainty in our estimates by analyzing the trace of uncertainty <span>$\delta \Theta^2 = \delta \Theta_1^2 + \delta \Theta_2^2 + \delta \Theta_3^2$</span> A scatter plot of the uncertainty versus average expression level is shown below. We see that expression level is an imperfect predictor of MLE uncertainty. The dashed cyan line denotes our chosen uncertainty cutoff used to remove genes from our count matrix. The right figure displays the cumulative density function for the filtered genes; as can be seen these are lowly expressed genes that are, in practice, 2 state variables.</p><p align="center">
<figure>
  <img src="/seqspace/assets/drosophila/nb_total_uncertainty_vs_expression.png" width="49%" />
  <img src="/seqspace/assets/drosophila/nb_badfits.png" width="49%" />
  <figurecaption>
  Filter genes with bad fits. Genes with high uncertainty are determined to be lowly expressed.
  </figurecaption>
</figure>
</p><p>Interestingly, as shown below, the mean estimated value for <span>$\Theta_2$</span> is <em>slightly higher</em> than 1. This phenomenon persists across all gene expression levels.</p><p align="center">
<figure>
  <img src="/seqspace/assets/drosophila/nb_param2.png" width="49%" />
  <img src="/seqspace/assets/drosophila/nb_param3.png" width="49%" />
  <figurecaption>
  Parameter distributions
  </figurecaption>
</figure>
</p><h4 id="Verify-estimates-via-bootstrap"><a class="docs-heading-anchor" href="#Verify-estimates-via-bootstrap">Verify estimates via bootstrap</a><a id="Verify-estimates-via-bootstrap-1"></a><a class="docs-heading-anchor-permalink" href="#Verify-estimates-via-bootstrap" title="Permalink"></a></h4><p>In order to test that the method does not overfit the data, especially lowly expressed genes, we validated our estimates via bootstrap. Specifically, we re-ran our maximum likelihood estimation of <span>$\{\Theta_{1;i}, \Theta_{2;i}, \Theta_{3;i} \}$</span> over subsamples of our given set of cells 100 times for each gene. The mean and standard deviation of the resultant empirical distribution was compared directly to our estimates and uncertainty calculations performed on the full dataset. As shown below, we find great quantitative agreement between the empirical estimates and our original calculations, suggesting strongly that we are <em>not</em> overfitting the data.</p><p align="center">
<figure>
  <img src="/seqspace/assets/drosophila/bootstrap_1.png" width="60%" />
  <img src="/seqspace/assets/drosophila/bootstrap_2.png" width="60%" />
  <img src="/seqspace/assets/drosophila/bootstrap_3.png" width="60%" />
</figure>
</p><h3 id="Drosophila-results"><a class="docs-heading-anchor" href="#Drosophila-results">Drosophila results</a><a id="Drosophila-results-1"></a><a class="docs-heading-anchor-permalink" href="#Drosophila-results" title="Permalink"></a></h3><p>Just as we performed using the toy data, once parameters <span>$\{\Theta_{1;i}, \Theta_{2;i}, \Theta_{3;i} \}$</span> are estimated, we can utilize Eqs. (2-5) to normalize the variance matrix and subsequently estimate the mean <span>$\mu_{i\alpha}$</span>. As shown below, we find there are <span>$\sim 30$</span> <strong>statistically significant</strong> linear dimensions in the scRNAseq data obtained during early <em>Drosophila melongaster</em> embryogenesis. Interestingly, while not fully delocalized as seen by the participation ratio of the &quot;noise&quot; components, we see that roughly <span>$\sim 1000$</span> genes contribute significantly to each component suggesting these are coarse &quot;pathways&quot; discovered.</p><p align="center">
<img src="/seqspace/assets/drosophila/rank_estimate.png" width="49%" class="center"/>
<img src="/seqspace/assets/drosophila/participation_ratio.png" width="49%" class="center"/>
</p><p>To ensure element-wise positivity, the estimated <span>$\tilde{\mu}_{i\alpha}$</span> is obtained by performing non-negative matrix factorization on the rescaled <span>$\tilde{n}_{i\alpha}$</span> with rank <span>$35$</span>. The factorization was initialized using the nndsvda algorithm <sup class="footnote-reference"><a id="citeref-6" href="#footnote-6">[6]</a></sup> and minimized the least squares objective via a multiplicative update <sup class="footnote-reference"><a id="citeref-7" href="#footnote-7">[7]</a></sup>.</p><h2 id="Pearson-residuals"><a class="docs-heading-anchor" href="#Pearson-residuals">Pearson residuals</a><a id="Pearson-residuals-1"></a><a class="docs-heading-anchor-permalink" href="#Pearson-residuals" title="Permalink"></a></h2><p>Once obtained, <span>$\tilde{\mu}_{i\alpha}$</span> provides us an estimate for the rescaled mean of the expression of gene <span>$i$</span> for cell <span>$\alpha$</span>. It is important to note, this <em>will</em> still have gene-specific and cell-specific scales by construction. However, our goal was originally to convert our raw count matrix into more natural units where variation across sequencing depth and gene expression are normalized out. As such, our normalization pipeline requires one last step: convert each rescaled mean into a z-score that measures expression differentially to the observed distribution across cells. Recall that a negative binomial stochastic model is equivalent to a Poisson-Gamma mixture, i.e. a poisson process whose mean is drawn from a Gamma distribution. To this end, we estimate the significance of each <span>$\tilde{\mu}_{i\alpha}$</span> by fitting a Gamma distribution across cells for each gene, in exactly the same way as we performed for the raw count data with a negative binomial distribution. Specifically, we take the mean to be given by Eq. (4) and minimize the analog of Eq. (5); we omit details here in the interest of brevity. This was determined to be an excellent stochastic model for the computed <span>$\tilde{mu}_{i\alpha}$</span>, as shown by the linear quantile-quantile plot below.</p><p align="center">
<img src="/seqspace/assets/drosophila/gamma_qq.png" width="49%" class="center"/>
</p><p>Taken together, our normalized gene count <span>$z_{i\alpha}$</span> is given by (the mean and variance is given by the estimated Gamma distribution)</p><p class="math-container">\[    \tag{6} z_{i\alpha} \equiv \frac{\tilde{\mu}_{i\alpha} - \mu_{i\alpha}}{\sigma_{i\alpha}}\]</p><section class="footnotes is-size-7"><ul><li class="footnote" id="footnote-1"><a class="tag is-link" href="#citeref-1">1</a><a href="https://arxiv.org/abs/1103.2221">The singular values and vectors of low rank perturbations of large rectangular random matrices</a></li><li class="footnote" id="footnote-2"><a class="tag is-link" href="#citeref-2">2</a><a href="https://arxiv.org/abs/math/0605624">The largest eigenvalue of rank one deformation of large Wigner matrices</a></li><li class="footnote" id="footnote-3"><a class="tag is-link" href="#citeref-3">3</a><a href="http://www3.stat.sinica.edu.tw/statistica/oldpdf/A17n418.pdf">Asymptotics of Sample Eigenstructure for a Large Dimensional Spiked Covariance Model</a></li><li class="footnote" id="footnote-4"><a class="tag is-link" href="#citeref-4">4</a><a href="https://arxiv.org/abs/2103.13840">Biwhitening Reveals the Rank of a Count Matrix</a></li><li class="footnote" id="footnote-5"><a class="tag is-link" href="#citeref-5">5</a><a href="https://arxiv.org/abs/1609.06349">A Review of Matrix Scaling and Sinkhorn&#39;s Normal Form for Matrices and Positive Maps</a></li><li class="footnote" id="footnote-6"><a class="tag is-link" href="#citeref-6">6</a><a href="https://www.sciencedirect.com/science/article/pii/S0031320307004359">SVD based initialization: A head start for nonnegative matrix factorization</a></li><li class="footnote" id="footnote-7"><a class="tag is-link" href="#citeref-7">7</a><a href="https://www.cs.cmu.edu/~11755/lectures/Lee_Seung_NMF.pdf">Algorithms for Non-negative Matrix Factorization</a></li></ul></section></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../">« Home</a><a class="docs-footer-nextpage" href="../inference/">scRNAseq spatial inference »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.12 on <span class="colophon-date" title="Tuesday 1 March 2022 15:54">Tuesday 1 March 2022</span>. Using Julia version 1.7.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
